### DISCUSSION TITLE
Rethink client transaction API

### DISCUSSION THREAD

**Discussion by 1st1** [geldata/gel#1708] [Ideas]
@tailhook suggested in Slack:

I'd like to have queries executed on the transaction object itself rather than on connection:

```python
with conn.transaction() as t:
    t.execute("xxx")
```

Or

```javascript
conn.transaction((t) => {
    t.execute("xxx")
})
```

The reasons are:
1. Passing transaction to helper functions explicitly (rather than connection which might have active transaction or might have not) looks better

2. This prevents things accidentally get into transaction, if connection is any kind of global and any function can call it.

3. We're likely do that in Rust anyway, as we need to borrow a connection for the transaction.

4. This allows pool.transaction() directly rather than two layers of with.

5. Basically, there is no need to use connection in 99% apps, after this change, it's either pool.query() or transaction.query(). So we get rid of the pattern of acquiring a connection for a long time and holding it (for a duration of HTTP request for example).
(and if we're going to fix this we should also do "repeating a transaction" by default too)

**Comment by 1st1** [comment]
I like the idea.

Few comments:

> This prevents things accidentally get into transaction, if connection is any kind of global and any function can call it.

Not sure how it would prevent "accidental" transactions and what they are.

> This allows pool.transaction() directly rather than two layers of with.

That's a nice improvement indeed.

> Basically, there is no need to use connection in 99% apps, after this change, it's either pool.query() or transaction.query(). So we get rid of the pattern of acquiring a connection for a long time and holding it (for a duration of HTTP request for example).

Yes, I can see that, provided we can have clear tutorials suggesting the pattern.

-----

There are also some modes of operation that are valid only within a transaction, like creating a cursor. If we implement this proposal, the `cursor()` method will only be implemented on the transaction object, gently nudging users towards using the db correctly.

One more idea: perhaps we can add a way to acquire a connection from the pool without `transaction()` method. And create a pool without `transaction()` and `acruire()` methods. That way you can ensure in one place that all connections are short lived.

**Reply by tailhook** [reply]
> > This prevents things accidentally get into transaction, if connection is any kind of global and any function can call it.
> 
> Not sure how it would prevent "accidental" transactions and what they are.
> 

Like this:
```python
connection = None

@cached
def site_info():
    return connection.query_row("SELECT SiteInfo {...}")

def request():
    global connection
    with pool.acquire() as conn:
        connection = conn
        with conn.transaction():
            site = site_info()
            do_something(site)
```
Here `SELECT SiteInfo` is sometimes a part of transaction, sometimes is not, depending on how lucky the request is to be the first one. While it should be an independent singleton that fetches and populates the cache.

While this extra select looks innocent, there can be writes like `INSERT .. UNLESS CONFLICT` or even more complex things.

Note, I don't think this is always wrong. Just have to be explicit. So if we have an abstract base class, say `Query`, that is implemented for all three: `Pool`, `Transaction`, `Connection`. And function that is not independent of current transaction can accept `do_domething(db: Query)`, and be used either with pool or transaction. And similar thing can be put into a global variable or context.

**Reply by 1st1** [reply]
While I think this example is a bit contrived, I see what you mean. Point taken.

**Comment by 1st1** [comment]
@tailhook 

> (and if we're going to fix this we should also do "repeating a transaction" by default too)

I'm not sure I like the idea of doing that by default. But can you elaborate a little on how you would do this?

**Reply by tailhook** [reply]
Basically this: https://www.cockroachlabs.com/docs/stable/transactions.html#client-side-intervention-example
Except, that is more pseudo code, so we also need at least:
1. Also retry on connection error (so that client reconnects gracefully in the middle of transaction, this by the way is another reason why we need **`transaction()` method on pool** rather than connection. 
2. Better error reporting than what's in example

**Reply by 1st1** [reply]
> Basically this: https://www.cockroachlabs.com/docs/stable/transactions.html#client-side-intervention-example

Yes, but IIRC you once proposed a more high-level API than what they show in their pseudo-code, which as is won't be used by anyone as it's too verbose.

**Reply by tailhook** [reply]
> > Basically this: https://www.cockroachlabs.com/docs/stable/transactions.html#client-side-intervention-example
> 
> Yes, but IIRC you once proposed a more high-level API than what they show in their pseudo-code, which as is won't be used by anyone as it's too verbose.

I don't remember anything like that. For common case API should look like current one in JavaScript (but documented that it may run function more than once). For Python, yes, it changes more, because Python doesn't have inline functions. We probably can use decoratorish syntax to make it simpler.

Sometimes tweaking transaction parameters are needed, though. That may be more complex, but that shouldn't be common.

**Reply by tailhook** [reply]
Continued in https://github.com/edgedb/edgedb/discussions/1738

**Comment by CodesInChaos** [comment]
One important note for the rust version: `drop` can't return an error. So committing a transaction needs to be an explicit call (which consumes the transaction), while dropping a transaction should roll it back (and swallow any errors happening during rollback).

**Reply by 1st1** [reply]
> (and swallow any errors happening during rollback).

Is it possible to not implement "drop" trait at all? Or make it a compile error if the transaction wasn't explicitly committed or rolled back?

**Reply by CodesInChaos** [reply]
Rust does not support linear types. Not implementing drop means the transaction would be dropped without side effects, which is certainly not what you want.

And even if all normal codepaths had an explicit rollback or commit, panics would still drop the transaction.

**Reply by 1st1** [reply]
> And even if all normal codepaths had an explicit rollback or commit, panics would still drop the transaction.

Yeah, panics dropping the actual transactions is clear. But is it possible to design an API where it's ensured at compile time that either `rollback()` or `commit()` is called explicitly on a transaction object?

**Reply by tailhook** [reply]
No, it's not possible as far as I know. What we can do (in async API) is send the message from the destructor, so that another task will clean up for us. Because futures are inherently droppable, by no means you can ensure that future completes.

In sync code, the only problem is logging the error somewhere, though.

**Reply by CodesInChaos** [reply]
You could check if the current thread is panicking in `drop` and panic if it isn't panicking yet. Not sure if that's a good idea.

1. This detects cases where the user forgets to commit/rollback during normal execution.
1. `drop` being called is only a bug if it happens during normal execution, not if it happens during a panic. So silently rolling back in this case is correct.
2. `drop` panicking when the thread is already panicking is an immediate abort of the whole process. So not panicking in this case is essential to avoid crashing the program.

**Reply by tailhook** [reply]
It would work with sync code, but with async code, any construction like:
```rust
future1.race(future2)
future1.timeout(Duration::from_secs(1))
```
Would drop future at runtime, there is no way to prevent that. Being unable to use timeouts is bad idea.

Also a lot of say HTTP handlers drop the future when request is cancelled. There is a way to prevent that: spawn another future and use a channel, but that's de-optimization. And clause above is still valid for the spawned future too.

**Reply by CodesInChaos** [reply]
Perhaps simply define a general Log/Warning callback in the edgedb configuration which is called for this and any other warning (e.g. the server sending a warning/error message). The default implementation would log it to stderr, but the user can provide another callback if they want.

**Reply by elprans** [reply]
Yes, issuing a warning should be enough.  This is similar to how Python warns you about coroutines you forgot to `await`.  By the way, the situation in Python is equivalent.  If you don't `commit()` explicitly, your transaction should get rolled back in `__del__` and you should get a warning.

**Reply by tailhook** [reply]
> Perhaps simply define a general Log/Warning callback in the edgedb configuration which is called for this and any other warning (e.g. the server sending a warning/error message). The default implementation would log it to stderr, but the user can provide another callback if they want.

Yes for sync case. For async case, warning doesn't work either because `.race` or `.timeout` or a lot of similar cases will drop future and user can't do anything with this. We can have some debug-level message, that can be turned on specifically for debugging this thing, but not more than that. Otherwise, users will complain that there are unfixable warnings and it's slow (because emitting a warning is slow, even when filtered out; in Rust's meaning of slowness of course).

**Reply by CodesInChaos** [reply]
I'd pass an enum to some kind of `notice_callback` function configured in the builder. The interpretation of the severity of specific events is up to the user. This should be pretty cheap, even by rust standards.

**Reply by tailhook** [reply]
Callback is fine performance-wise but:
1. I'm not sure anybody will set that. So default should be sensible.
2. I'm not sure this is "everything that is notice-level", we may want an enum, similar to error handling spec (or that's too much)
3. I'm not sure about location tracking. How to find out what exact transaction is not closed. Should we expect user to call `Backtrace::new()` and do whatever they want? We probably cant use `track_caller` there. Unless we want to give names to transactions.

**Reply by tailhook** [reply]
>     2. I'm not sure this is "everything that is notice-level", we may want an enum, similar to error handling spec (or that's too much)
> 

More context: we probably will need another notice/warning when transaction was repeated (if we implement that as a builtin). And we should give them a way to differentiate between `transaction.rollback()` is skipped and transaction was repeated without matching the message. This should be useful for gathering run-time statistics.

**Reply by CodesInChaos** [reply]
I'd pass in a mix of general parameters (e.g. the connection), and event specific parameters.  Probably something like:

```
struct NoticeContext<'t> {
    event:NoticeEvent,
}

struct TransactionDroppedEvent<'t> {
    transaction:&Transaction,
}

enum NoticeEvent<'t> {
    TransactionDropped(TransactionDroppedEvent),
    ...
}

fn notice_callback(self:Builder, impl Fn(&NoticeContext));
```

If the user wants a backtrace they can create it themselves. The default would probably just print to stderr. Personally I'd probably use some variant of the tracing crate, and use its spans to figure out which context the transaction was associated with.

**Reply by tailhook** [reply]
Yes something like this. The problem is that transaction object per se doesn't mean anything, i.e. there is nothing that could differentiate one transaction object from another saying which part of the code uses it or whatever.

Is there any good tracing lib that works with async? We should probably look into how to integrate that.

**Reply by CodesInChaos** [reply]
> Is there any good tracing lib that works with async?
From what I read, the `tracing` crate (or one of its support crates) can link `event`s to the surrounding `span`, even in async code. I did not try it myself though. Not sure if this relies on tokio or also works with async-std.

> We should probably look into how to integrate that.

I'm not a fan of low level libraries integrating with a logging framework. I prefer it providing the necessary callbacks which the user can link up with the logging framework.

**Reply by tailhook** [reply]
> I'm not a fan of low level libraries integrating with a logging framework. I prefer it providing the necessary callbacks which the user can link up with the logging framework.

Sure, most of the time yes. But it's unclear what needs to be done. Even if nothing, we might want to give example in the docs for usage of such libraries, because it doesn't look like a trivial task to identify transaction in a callback.

**Comment by tailhook** [comment]
We also should think how sub-transactions should be done:
```
with transaction.savepoint('xxx') as savepoint:
    savepoint.query('...')
```
I.e. I would temporary block respective methods in `Transaction` too.

**Reply by elprans** [reply]
Hm. Savepoints aren't really sub-transactions.  You can't "commit" a savepoint, and savepoints may overlap and get released or rolled-back to arbitrarily.  You _can_ emulate a nested transaction with savepoints, but we probably shouldn't be using the `savepoint` name for this.

**Reply by 1st1** [reply]
> You can emulate a nested transaction with savepoints

That's precisely what we're doing with our API to make it composable. I don't like the idea of exposing "savepoint" terminology in a high-level API. We *could* call "nested" transactions something else, like "blocks" to make it explicit what's really going on there.
